homeassistant/helpers/config_entry_oauth2_flow.py

```diff
@@ -329,17 +329,29 @@ def async_register_implementation(
 @callback
 def async_register_implementation(
     hass: HomeAssistant, domain: str, implementation: AbstractOAuth2Implementation
-) -> None:
+) -> Callable[[], None]:
     """Register an OAuth2 flow implementation for an integration."""
-    if isinstance(implementation, LocalOAuth2Implementation) and not hass.data.get(
-        DATA_VIEW_REGISTERED, False
-    ):
-        hass.http.register_view(OAuth2AuthorizeCallbackView())
-        hass.data[DATA_VIEW_REGISTERED] = True
+    if isinstance(implementation, LocalOAuth2Implementation):
+        async_register_local_apis(hass)
 
     implementations = hass.data.setdefault(DATA_IMPLEMENTATIONS, {})
     implementations.setdefault(domain, {})[implementation.domain] = implementation
 
+    def unsub() -> None:
+        del implementations[domain][implementation.domain]
+
+    return unsub
+
+
+def async_register_local_apis(hass: HomeAssistant) -> None:
+    """Initialize requirements for local OAuth2 implementations.
+
+    This must be called on startup before the http router is frozen.
```

Comment:
We prevent the router from getting frozen so this is not an issue to work around.

Reply:
I am having trouble reconciling your comment with the error message I see:
```
homeassistant/data_entry_flow.py:205: in async_init
    flow, result = await task
homeassistant/data_entry_flow.py:232: in _async_init
    result = await self._async_handle_step(flow, flow.init_step, data, init_done)
homeassistant/data_entry_flow.py:335: in _async_handle_step
    result: FlowResult = await getattr(flow, method)(user_input)
homeassistant/helpers/config_entry_oauth2_flow.py:319: in async_step_user
    return await self.async_step_pick_implementation(user_input)
homeassistant/helpers/config_entry_oauth2_flow.py:236: in async_step_pick_implementation
    implementations = await async_get_implementations(self.hass, self.DOMAIN)
homeassistant/helpers/config_entry_oauth2_flow.py:370: in async_get_implementations
    async_register_local_apis(hass)
homeassistant/helpers/config_entry_oauth2_flow.py:349: in async_register_local_apis
    hass.http.register_view(OAuth2AuthorizeCallbackView())
homeassistant/components/http/__init__.py:295: in register_view
    view.register(self.app, self.app.router)
homeassistant/components/http/view.py:95: in register
    routes.append(router.add_route(method, url, handler))
venv/lib/python3.9/site-packages/aiohttp/web_urldispatcher.py:1100: in add_route
    resource = self.add_resource(path, name=name)
venv/lib/python3.9/site-packages/aiohttp/web_urldispatcher.py:1085: in add_resource
    self.register_resource(resource)

...
resource = <PlainResource  /auth/external/callback>

    def register_resource(self, resource: AbstractResource) -> None:
        assert isinstance(
            resource, AbstractResource
        ), f"Instance of AbstractResource class is required, got {resource!r}"
        if self.frozen:
>           raise RuntimeError("Cannot register a resource into frozen router.")
E           RuntimeError: Cannot register a resource into frozen router.
```

My impression is that we are in fact freezing the router and this registration must be done at startup time, rather than just in time when the first local oauth implementation is registered.  Should I interpret your comment to mean the test harness is incorrect?

Reply:
The above error happens if i do the API registration lazily, like how the non-provider does it for local oauth implementations.  An alternative framing is that the integration just needs to call this to have the APIs exposed at all when using the provider interface, so i removed the comment.

Reply:
https://github.com/home-assistant/core/blob/dev/homeassistant/components/http/__init__.py#L441-L446

Reply:
Well .. that nice "creative" :) solution to avoid freezing does not run during tests. I am not going to dig into this, as moving the registration to the a component seems like the best fix. I'll revert changes to this file in this PR.

Reply:
Yeah this is one of those ticking time bombs that will one aiohttp update explode in our face. I think that we can workaround it using a resource, [like we did for the frontend index](https://github.com/home-assistant/core/blob/dev/homeassistant/components/frontend/__init__.py#L503), but never dug in to solve it.